components/ca-crud.js

```diff
@@ -0,0 +1,810 @@
+        set showAll(value) {
+            this.setAttribute('show-all', value === true);
+        }
+
+        /**
+         * Loads the .src value via Ajax - can be either a schema/data url.
+         * If the request returns a JSON schema, it assigns the .schema property and checks for HATEOAS instances link of which it procecedes to load data
+         * If the request returns a JSON list of data, it assigns the .data property and checks for a HATEOAS describedby link of which is proceedes to load the schema
+         * @access private
+         * @returns {Promise} promise resolved when s
+         */
+        loadSrc() {
+
+            let schemaUrl = this.src;
+
+            return ajax.execute({ url: schemaUrl, dataType: 'json' }).then(data => {
+
+                if (Array.isArray(data.type)) {
+                    this.schema = data.type;
+                    this.instances = data.type.map(item => item.links.find(l => l.rel === 'instances'));
+                    this.loadData();
+                } else if (data.$schema) {
+                    this.schema = [data];
+                    this.instances = (data.links || []).filter(l => l.rel === 'instances');
+                    this.loadData();
+                } else if (data.list) {
+                    this.data = [data];
+                    this.instances = [{ href: schemaUrl, rel: 'instances' }];
+                    schemaUrl = data.links.find(l => l.rel === 'describedby').href;
+                    return ajax.execute({ url: schemaUrl, dataType: 'json' }).then(schema => {
+                        this.schema = [schema];
+                        this.render();
+                    });
+                }
+
+                return data;
+            });
+        }
+
+        /**
+         * Loads the data referenced by the instances href property. Executed when this.schema property changes
+         * @access private
+         * @returns {undefined} nothing
+         */
+        loadData() {
+
+            // convert each instance end point into a promise for loading
+            const promises = distinct(this.instances, item => item.href.toLowerCase()).map(item => {
+                if (item.schema && item.schema.properties && Object.keys(item.schema.properties).length) {
+                    // This URL is parameterized, so we can't load it in advance.
+                    return Promise.resolve([]);
+                }
+                return ajax.execute({ url: item.href, dataType: 'json' });
+            });
+
+            // load all promises and then merge the data
+            Promise.all(promises).then(data => {
+                this.data = Array.isArray(data) ? data : [data];
+                this.render();
+            }).catch(err => {
+                trace(err);
+            });
+        }
+
+        /**
+         * Simply reloads the data
+         * @access private
+         * @returns {undefined} nothing
+         */
+        reload() {
+            this.loadData();
```

Comment:
Why not just call this.loadData()  where required?  Why is this wrapped in a function?